atc/db/resource_check_rate_limiter.go

```diff
@@ -0,0 +1,105 @@
+package db
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"code.cloudfoundry.org/clock"
+	"golang.org/x/time/rate"
+)
+
+type ResourceCheckRateLimiter struct {
+	clock clock.Clock
+
+	checkInterval time.Duration
+	checkLimiter  *rate.Limiter
+
+	refreshConn    Conn
+	refreshLimiter *rate.Limiter
+
+	mut *sync.Mutex
+}
+
+func NewResourceCheckRateLimiter(
+	refreshConn Conn,
+	checkInterval time.Duration,
+	refreshInterval time.Duration,
+	clock clock.Clock,
+) *ResourceCheckRateLimiter {
+	return &ResourceCheckRateLimiter{
+		clock: clock,
+
+		checkInterval: checkInterval,
+		checkLimiter:  nil,
+
+		refreshConn:    refreshConn,
+		refreshLimiter: rate.NewLimiter(rate.Every(refreshInterval), 1),
+
+		mut: new(sync.Mutex),
+	}
+}
+
+func (limiter *ResourceCheckRateLimiter) Wait(ctx context.Context) error {
+	limiter.mut.Lock()
+	defer limiter.mut.Unlock()
+
+	if limiter.refreshLimiter.AllowN(limiter.clock.Now(), 1) {
+		err := limiter.refreshCheckLimiter()
+		if err != nil {
+			return fmt.Errorf("refresh: %w", err)
+		}
+	}
+
+	reservation := limiter.checkLimiter.ReserveN(limiter.clock.Now(), 1)
+
+	delay := reservation.DelayFrom(limiter.clock.Now())
+	if delay == 0 {
+		return nil
+	}
+
+	timer := limiter.clock.NewTimer(delay)
+	defer timer.Stop()
+
+	select {
+	case <-timer.C():
+		return nil
+	case <-ctx.Done():
+		reservation.Cancel()
+		return ctx.Err()
+	}
+}
+
+func (limiter *ResourceCheckRateLimiter) Limit() rate.Limit {
+	limiter.mut.Lock()
+	defer limiter.mut.Unlock()
+
+	return limiter.checkLimiter.Limit()
+}
+
+func (limiter *ResourceCheckRateLimiter) refreshCheckLimiter() error {
+	var count int
+	err := psql.Select("COUNT(id)").
+		From("resource_config_scopes").
+		RunWith(limiter.refreshConn).
+		QueryRow().
+		Scan(&count)
+	if err != nil {
+		return err
+	}
+
+	limit := rate.Limit(float64(count) / limiter.checkInterval.Seconds())
```

Comment:
Should this take into account `--max-checks-per-second` as we used to do? If that's specified, we don't even need to count the resource config scopes/refresh the limit - so maybe it'd be better as a separate implementation of `builder.RateLimiter` (e.g. `StaticCheckRateLimiter`)